### ISSUE TITLE
Checks / cleanup before allocation of unmanaged regions

### ISSUE THREAD

**Issue by sy-c** [FairRootGroup/FairMQ#367]
**Is your feature request related to a problem? Please describe.**

We have instabilities on O2 FLPs when there are left-over of unmanaged memory regions. Restarting can crash the machine if there is not enough memory left when trying to use a fresh big new block given by FMQ library.
Our experience shows that in some cases cleaning "after" is not enough, and it would help to do it "before" too - and at the same time check that the new memory block requested can actually be used.

**Describe the solution you'd like**

Adding option(s) to CreateUnmanagedRegion() so that it performs the 3 steps below:
1) Cleanup existing leftover in /dev/shm
  - can check and remove only inactive files (e.g. as seen with one of fuser /dev/shm/*; stat -c %Y /dev/shm/*; lsof +D /dev/shm )
  - possibly only those shm files related to the uuid of the matching transport, to avoid any interference with other regions
2) Check if currently available free memory can accommodate the new block, and abort if not. (/proc/meminfo MemFree)
3) Commit the memory immediately to avoid bad surprises later (although writing all takes a long time... 1.65 GB/s typically on our reference system)

**Describe alternatives you've considered**
This could be done in the control system or in the client using the FMQ API, but would not provide the best coverage compared to executed in the library directly at time of region creation. It makes sense that this useful functionality belongs to the library.

**Comment by rbx** [comment]
> Our experience shows that in some cases cleaning "after" is not enough

Can you elaborate? If there is a bug with the cleaning, it needs to be fixed. Or what do you mean cleaning after is not enough?

>     1. Cleanup existing leftover in /dev/shm

Within the same session, consecutively created regions will result in different file names. Basically, per session there is an internal counter, that also resides in the session shmem, which is incremented with each unmanaged region. So a request for a new UR will increment the counter first and then try to get the memory. Meaning, a request for a new region (1) will not try to access the same file and (2) would not know which other region to cleanup if it wanted to do so. The Cleanup API is there for "emergency cases" (e.g. when something has crashed) and removes everything that belongs to a session. More granular cases should be covered by an orderly object destruction.

>     1. Check if currently available free memory can accommodate the new block, and abort if not. (/proc/meminfo MemFree)

We considered this in the past, and basically our conclusion was that there can be cases where one might want to allocate more memory than is actually available, e.g. if the access pattern avoids parts of it. But perhaps this can be an option. @dennisklein thoughts?

>     2. Commit the memory immediately to avoid bad surprises later (although writing all takes a long time... 1.65 GB/s typically on our reference system)

There is now an option to mlock and/or zero the created memory before the ctor returns: https://github.com/FairRootGroup/FairMQ/blob/master/examples/region/sampler.cxx#L59
Would this option suffice for what you have in mind?

**Comment by dennisklein** [comment]
> > ```
> > 1. Check if currently available free memory can accommodate the new block, and abort if not. (/proc/meminfo MemFree)
> > ```
> 
> We considered this in the past, and basically our conclusion was that there can be cases where one might want to allocate more memory than is actually available, e.g. if the access pattern avoids parts of it. But perhaps this can be an option. @dennisklein thoughts?

What internal knowledge of FairMQ is required to justify such a logic being built-in? Why not just compose this functionality naturally (pseudo-code):

```
if (GetFreeMemory() >= size) {
  // create the region with size
} else {
  // throw or whatever
}
```

**Comment by sy-c** [comment]
Hello, and thanks for your feedback and taking time to consider this request!

> > Our experience shows that in some cases cleaning "after" is not enough
> Can you elaborate? If there is a bug with the cleaning, it needs to be fixed. Or what do you mean cleaning after is not enough?

Despite using the shm cleanup tool - which is already helping a lot - there are still some rare situations when the state is not clean on restarting (after something really bad happened). We have been reported a number of cases in the past days. I leave @teo and @vascobarroso comment on what is done currently in the O2 control.
My statement is a bit more general (and even not specifically linked to this problem). Even if there are good cleanup procedures after stop which probably cover most of the cases, there are still a few rare cases when we want to ensure everything is really clean just before starting. So it's not necessarily about the existing "after stop" cleanup procedure having a bug, it's about adding a (more radical and optional) "before start" cleanup procedure.


> . Cleanup existing leftover in /dev/shm

I was possibly considering something more drastic than "per session", like removing all regions not used by any process.


> What internal knowledge of FairMQ is required to justify such a logic being built-in? Why not just compose this functionality naturally

When using malloc() or hugepages to get big memory blocks for the readout software, the corresponding APIs would immediately return an error if requesting too much, and allow a graceful exit. When using FairMQ unmanaged region, no error is reported and there are later uncontrolled crashes when using the memory. I think it makes sense to (optionally) have the same kind of feature builtin as other memory supports to ensure that what is returned is a fully usable object, and to detect issues before they happen.


> There is now an option to mlock and/or zero the created memory before the ctor returns: https://github.com/FairRootGroup/FairMQ/blob/master/examples/region/sampler.cxx#L59
> Would this option suffice for what you have in mind?

This looks exactly what I need for this point :)
It's just a pity that it also crashes the process if not enough memory is available.
I notice in the same code the use of a "setLinger()" call... does it relate to the region cleanup after exit?


cheers,
Sylvain

**Comment by rbx** [comment]
> My statement is a bit more general (and even not specifically linked to this problem). Even if there are good cleanup procedures after stop which probably cover most of the cases, there are still a few rare cases when we want to ensure everything is really clean just before starting. So it's not necessarily about the existing "after stop" cleanup procedure having a bug, it's about adding a (more radical and optional) "before start" cleanup procedure.

> I was possibly considering something more drastic than "per session", like removing all regions not used by any process.

Because of reasons outlined in #330 I would not want to add a "delete all" call at all. And even if we would go for it, calling it during region creation (that belongs to a specific session) would be wrong. Finally, in the end it would just call the same API as is called during the shutdown or by the current cleanup tool - so if that has already failed for some reason, calling it would just likely fail again. So the next "escalation step" would be to just `rm` the files as sudo.

> I notice in the same code the use of a "setLinger()" call... does it relate to the region cleanup after exit?
Not really. UnmanagedRegion sets up a queue (at least in the shmem case) to receive events about buffers that are no longer needed. The linger option is there to add a delay during destruction, to try and receive the events that have not arrived yet. When this delay expires, region will be deleted, no matter how many events are still outstanding.

**Comment by dennisklein** [comment]
> > What internal knowledge of FairMQ is required to justify such a logic being built-in? Why not just compose this functionality naturally
>
> When using malloc() or hugepages to get big memory blocks for the readout software, the corresponding APIs would immediately return an error if requesting too much, and allow a graceful exit.

I cannot follow, for me a successfull `malloc` does not reflect availability of physical memory pages.

> When using FairMQ unmanaged region, no error is reported and there are later uncontrolled crashes when using the memory.

The idea of the FairMQ unmanaged region interface is to "register" some virtual memory region with FairMQ and to handle msg buffer ownership transfers, but leave the management of msg buffers within this memory region to the user code (hence the name *unmanaged* (from the point of view of FairMQ)). "uncontrolled crashes" sounds to me like the management entity of this region is lacking some error handling.

> I think it makes sense to (optionally) have the same kind of feature builtin as other memory supports to ensure that what is returned is a fully usable object, and to detect issues before they happen.

With this formulation I agree ;), but you proposed

> Check if currently available free memory can accommodate the new block, and abort if not. (/proc/meminfo MemFree)

What would even be the meaning of checking availability of free physical pages at the creation time of a virtual memory region? It would not guarantee anything, because the moment after the check, any arbitrary system component (e.g. completely unrelated processes) could use up the physical pages, right?

IMHO, the only meaningful thing for usage safety would be to *pin* the region (as in directly allocate physical memory) (right now we have `zero`ing and `mlock`ing support, I am open to add any other useful option there). On the other hand, any more granular *pin*ing can probably also be done by the user memory manager managing this region. One would have to check the concrete case, there are lots of techniques, some may require some interaction with FairMQ, others may not.

**Comment by sy-c** [comment]
> > When using malloc() or hugepages to get big memory blocks for the readout software, the corresponding APIs would immediately return an error if requesting too much, and allow a graceful exit.
> I cannot follow, for me a successfull `malloc` does not reflect availability of physical memory pages.

malloc() indeed does not guarantee they are available at the time you use them, but there are still some basic checks done at the time malloc() is called. Of course, if by the time you write to it there's not enough space, OOM comes into the play and kills processes around, usually the biggest consumers (which can be something else than the caller).

The tests I have done in the past days on CentOS 7 confirm that if trying to malloc() more than currenlty free memory size, it returns a nullptr and errno 12 : Cannot allocate memory

I don't ask more than that for FairMQ. At the moment, it always end up in a bus error (if /dev/shm full) or a process kill (if no memory left). Which in turn usually leaves behind some left-over in /dev/shm.
Having an (optional) basic check for these two obviously dangerous conditions, with a graceful error at the time of region creation (especially when using the mlock or zero options) would help a lot to prevent going in a bad state. (when such over-committed memory is used for a device driver, we have seen situations stuck at the point where nothing but a reboot would help).

I can certainly implement such protection on top of FairMQ in my own client, but I believe it's a valuable add-on for other FairMQ users as well. I would for sure appreciate if you decide to add this feature built-in, but also understand that you may have some reasons not to do so. Please let me know!

**Comment by dennisklein** [comment]
> > > When using malloc() or hugepages to get big memory blocks for the readout software, the corresponding APIs would immediately return an error if requesting too much, and allow a graceful exit.
> >
> > I cannot follow, for me a successfull `malloc` does not reflect availability of physical memory pages.
>
> malloc() indeed does not guarantee they are available at the time you use them, but there are still some basic checks done at the time malloc() is called. Of course, if by the time you write to it there's not enough space, OOM comes into the play and kills processes around, usually the biggest consumers (which can be something else than the caller).
> 
> The tests I have done in the past days on CentOS 7 confirm that if trying to malloc() more than currenlty free memory size, it returns a nullptr and errno 12 : Cannot allocate memory

:thinking: Then your `malloc()` implementation is different from mine.

<details>
<summary>I can allocate ~37GiB in an individual <tt>malloc()</tt> call (appears to be total physical RAM plus total Swap?) and that 3541 times (just short of the linux x86_64 <a href="https://www.kernel.org/doc/Documentation/x86/x86_64/mm.txt">128 TiB VM limit</a>) on my notebook with 16GB physical RAM:</summary>

```console
$ free -h
               total        used        free      shared  buff/cache   available
Mem:            15Gi       6.8Gi       1.3Gi       1.2Gi       7.4Gi       7.1Gi
Swap:           21Gi       3.2Gi        18Gi
$ cat malloc.cpp
#include <cstdlib>
#include <iostream>

int main(int argc, char *argv[])
{
  for(int i = 0; i < 10000; ++i) {
    if(malloc((size_t)1024*1024*1024*37) == nullptr) {
      std::cout << i << std::endl;
      break;
    }
  }

  (void)std::cin.get();
  return 0;
}
$ g++ malloc.cpp -o malloc
$ ./malloc
3541

```
```console
$ ps -A | grep malloc
3628675 pts/19   00:00:00 malloc
$ pmap 3628675 | grep total
total     137381301932K
```
What am I missing?
</details>

> I can certainly implement such protection on top of FairMQ in my own client, but I believe it's a valuable add-on for other FairMQ users as well. I would for sure appreciate if you decide to add this feature built-in, but also understand that you may have some reasons not to do so. Please let me know!

I am still not convinced about any usefulness and/or effectiveness of the such a proposed *protection* in the `FairMQUnmanagedRegion` abstraction. Let's assume for a moment, it was effective(, because a region can magically? assume noone else will eat its physical memory pages). Against which limit would you check? - current free physical memory, ok, then I instantiate a second region without having written to the first region. free physical memory is still the same, so we would need to add shared state between all regions to track already "potentially allocated" physical memory... and this would need to be distributed shared state between all devices within a session (what about another session is spawned in parallel?). Is it what you had in mind?

**Comment by sy-c** [comment]
All right, let's forget about this. I'll implement it on my side and send it to you if you want to have a look.

I have a final request then: can you please raise an exception when the region is created with fair::mq::RegionConfig{true,...} option to lock the memory, but that it fails?
At the moment, an error is printed on output (so the fault is detected), but the function call returns normally, and the process crashes later when accessing the memory in user code.

[ERROR] Could not lock region 1. Code: 12, reason: Cannot allocate memory
Bus error (core dumped)

**Comment by sy-c** [comment]
thanks!

**Comment by rbx** [comment]
An exception on failing memory lock is included in the latest tag (it will throw `fair::mq::TransportError`). Closing this, as I don't see anything else actionable at this point.
